bitpoll/caldav/forms.py

```
@@ -0,0 +1,128 @@
+    TextInput
+from django.utils.translation import ugettext as _
+from urllib3.util import Url, parse_url
+
+from django_token_bucket.models import TokenBucket
+from .models import DavCalendar
+from ..base.models import BitpollUser
+
+
+class URLAuthWidget(MultiWidget):
+    """
+    A widget that splits datetime input into two <input type="text"> boxes.
+    """
+    template_name = 'caldav/url_auth.html'
+
+    def __init__(self):
+        widgets = (
+            URLInput(
+            ),
+            TextInput(
+            ),
+            PasswordInput(
+            )
+        )
+        super().__init__(widgets)
+
+    def decompress(self, value):
+        if value:
+            parsed = parse_url(value)
+            scheme, auth, host, port, path, query, fragment = parsed
+            url = u""
+            user = None
+            passwd = None
+
+            # We use "is not None" we want things to happen with empty strings (or 0 port)
+            if scheme is not None:
+                url += scheme + u"://"
+            if auth is not None:
+                user = auth.split(':')[0]
+                if ":" in auth:
+                    passwd = auth.split(':')[1]
+            if host is not None:
+                url += host
+            if port is not None:
+                url += u":" + str(port)
+            if path is not None:
+                url += path
+            if query is not None:
+                url += u"?" + query
+            if fragment is not None:
+                url += u"#" + fragment
+            return [url, user, passwd]
+        return [None, None, None]
+
+
+class URLAuthField(MultiValueField):
+    widget = URLAuthWidget
+
+    def __init__(self, **kwargs):
+        max_lengh = kwargs.pop('max_length')  # TODO use this in validation of the whole field
+        required = kwargs.pop('required')  # TODO use this in validation of the whole field
```

Would it not be better to instead raise an assertion error if one of these kwargs is set? (Or are they automatically set when populated from the model?) After all, we should not be setting these values anywhere ourselves.

---

This is set by the ModelForm because the database field is required and has a length